### PR TITLE
Use Content-Length to read the request body if available

### PR DESCRIPTION
**Pull request by casperisfine**
It allows the underlying server to better optimize the read.

### CODE REVIEW

File: lib/rack/utf8_sanitizer.rb [hunk 1]
```diff
@@ -125,7 +125,10 @@ def sanitize_rack_input(env)
       uri_encoded = URI_ENCODED_CONTENT_TYPES.any? {|type| content_type == type}
 
       if env['rack.input']
-        sanitized_input = sanitize_io(env['rack.input'], uri_encoded)
+        content_length = if env['CONTENT_LENGTH']
+          env['CONTENT_LENGTH'].to_i
+        end
+        sanitized_input = sanitize_io(env['rack.input'], uri_encoded, content_length)
```
Comment (Morriar) [inline lib/rack/utf8_sanitizer.rb:131]:
```suggestion
        sanitized_input = sanitize_io(env['rack.input'], uri_encoded: uri_encoded, content_length: env['CONTENT_LENGTH'].to_i)
```

File: lib/rack/utf8_sanitizer.rb [hunk 2]
```diff
@@ -167,8 +170,12 @@ def close
       end
     end
 
-    def sanitize_io(io, uri_encoded = false)
-      input = io.read
+    def sanitize_io(io, uri_encoded = false, content_length = nil)
+      input = if content_length
+        io.read(content_length)
```
Comment (Morriar) [inline lib/rack/utf8_sanitizer.rb:175]: If there any risk to crash/block if the `CONTENT_LENGTH` sent by the client is negative?

**Review by Morriar — APPROVED** [review]

File: lib/rack/utf8_sanitizer.rb (hunk 1 above)
Comment (casperisfine) [reply lib/rack/utf8_sanitizer.rb:131]: No we can't do that as `CONTENT_LENGTH` may not be present (chunked), so `nil.to_i` would return `0`

File: lib/rack/utf8_sanitizer.rb (hunk 2 above)
Comment (casperisfine) [reply lib/rack/utf8_sanitizer.rb:175]: That is dependent on the underlying server. But given that Rails will read the request body by passing `content_length`, I believe it's already handled by servers.

https://github.com/rails/rails/blob/939fd3bc41bbb90194e4b93a001c803bdceb592c/actionpack/lib/action_dispatch/http/request.rb#L467C14-L470

Comment (AlexWayfer) [reply lib/rack/utf8_sanitizer.rb:175]: As I know, you can send any value for any header. Yeah, it'd be out of standards, but it's still possible, and someone can use it (for example, as HTTP 200 OK with "not found" error).

Comment (casperisfine) [reply lib/rack/utf8_sanitizer.rb:175]: @AlexWayfer sorry I'm not sure what point you are trying to make.

Yes clients can send garbage in `Content-Length`, but that's no reason not to use it.

Comment (AlexWayfer) [reply lib/rack/utf8_sanitizer.rb:175]: As I understood, we're not talking about using it or not, we're discussing negative (or other) values which potentially could crash `io.read` code.

BTW, I see no tests changes in this PR, and this is… dangerous.

Comment (casperisfine) [reply lib/rack/utf8_sanitizer.rb:175]: Servers do handle negative Content-Length already.

> I see no tests changes in this PR

This codepath is already tested.

Comment (AlexWayfer) [reply lib/rack/utf8_sanitizer.rb:175]: > Servers do handle negative Content-Length already.

Which servers? Any servers? Thin, Puma, Unicorn, Falcon — any of them? I don't think we can rely this way. If Rack would handle, on which we rely (and depend) — then OK.

> This codepath is already tested.

If you introduce changes to the code — tests should be added or changed too.

For example, the old code always did `io.read`. Now we have condition — great, now we have 2 cases and both should be tested, truthy and falsey branches. Especially that `io` can contain more than `content_length`, but only `content_length` would be read, not the whole `io`.

Comment (casperisfine) [reply lib/rack/utf8_sanitizer.rb:175]: > Any servers? Thin, Puma, Unicorn, Falcon — any of them?

Yes. `rack.input` in Puma is either `StringIO` or `Tempfile`, both of which raise `ArgumentError` on negative size like Ruby's `IO`.

Unicorn/Pitchfork both explicitly raise argument error on such case too to mimick Ruby IO.

Falcon returns empty string on negative length.

Thin's `rack.input` is a `StringIO`, so raises as well.

Now the Rack spec says it's invalid to read a negative length, so I guess we can check for it.

> If you introduce changes to the code — tests should be added or changed too.

Not if the change is functionally equivalent.

File: lib/rack/utf8_sanitizer.rb (hunk 1 above)
Comment (casperisfine) [reply lib/rack/utf8_sanitizer.rb:131]: Yeah, but the gemspec claim to support 1.9.3 (thought CI only test 2.5+), and safe navigation was only added in 2.3. So I can't use it.

Comment (whitequark) [reply lib/rack/utf8_sanitizer.rb:131]: We can drop everything before 2.3.

Comment (casperisfine) [reply lib/rack/utf8_sanitizer.rb]: > We can drop everything before 2.3.

Done.

**Pull request merged**